File: Data Structures & Algorithms/two-integer-sum/submission-5.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> d;
        for (int i=0; i<nums.size(); i++) {
            if (d.contains(target - nums[i])) {
                return {d[target - nums[i]], i};
            }
            d[nums[i]] = i;
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-5.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(cur, total, pos):
            if total == target:
                res.append(cur.copy())
                return
            if pos >= len(nums) or total > target:
                return

            cur.append(nums[pos])
            dfs(cur, total + nums[pos], pos)
            cur.pop()
            dfs(cur, total, pos + 1)

        dfs([], 0, 0)
        return res